### PR TITLE
wrappers: support per-service quotas when writing service units (2/4)

### PR DESCRIPTION
**Pull request by Meulengracht**
This adds support for the service code to take into account that a sub-group might exist, which contains quotas for the specific service instead of using the quotas for the snap.

### CODE REVIEW

File: wrappers/services.go [hunk 1]
```diff
@@ -582,9 +582,42 @@ func (es *ensureSnapServicesContext) reloadModified() error {
...
+		}
+	}
+	return nil
+}
+
+// resolveAppSpecificQuotaGroup takes into account any service-specific quota group. Quota groups
+// are assigned on snap basis in general, which means the quota applies to everything inside the snap, unless
+// a sub-group exists that refer to a specific service. This function checks for that, and returns the correct
+// quota group, should a sub-group exist for this specific appInfo.
+func resolveAppSpecificQuotaGroup(appInfo *snap.AppInfo, snapQuotaGroup *quota.Group, allGrps []*quota.Group) *quota.Group {
```
Comment (pedronis) [inline wrappers/services.go:599]: two things, I think we have a check that we don't put the same snap in more than one group? do we have the same check for snap services, that they don't end up in more than one group? we should

File: wrappers/services.go [hunk 2]
```diff
@@ -582,9 +582,42 @@ func (es *ensureSnapServicesContext) reloadModified() error {
...
+	if snapQuotaGroup == nil {
+		return nil
+	}
+
+	for _, name := range snapQuotaGroup.SubGroups {
+		subGroup := quotaGroupByName(name, allGrps)
```
Comment (pedronis) [inline wrappers/services.go:605]: if we need to iterate over all the groups anyway I would do something else, intestead of going over the subgroups, I would go over all the groups check in the parent is snapQuotaGroup and then check if the app is in this group then. This a makes us iterate over all the groups only once

File: wrappers/services.go [hunk 3]
```diff
@@ -615,20 +648,27 @@ func (es *ensureSnapServicesContext) ensureSnapSystemdUnits(snapInfo *snap.Info,
 	}
 
 	// note that the Preseeding option is not used here at all
+	allGrps := quotaGroups.AllQuotaGroups()
 	for _, app := range snapInfo.Apps {
 		if !app.IsService() {
 			continue
 		}
```
Comment (pedronis) [inline wrappers/services.go:655]: we could also split this into two loops, first finding all the services, and have a function that does the resolving for all the apps in one go and returns a mapping app-name -> service quota group or nil which means use opts.Quota group then have a looping doing the actual generation

**Review by pedronis — COMMENTED** [review]
thanks, some suggestions to loop a bit less and a question about valid group setups checking

File: wrappers/services.go (hunk 1 above)
Comment (Meulengracht) [reply wrappers/services.go:599]: I agree with your comments on the approach, I'll refactor the loop approach a bit

**Comment by Meulengracht** [comment]
Thank you for the reviews here @pedronis . This PR will be superseeded by https://github.com/snapcore/snapd/pull/12395

**Pull request closed without merging**